osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs

```diff
@@ -109,33 +151,48 @@ private void load()
                 };
 
                 bank.TabbableContentContainer = flow;
+                additionBank.TabbableContentContainer = flow;
                 volume.TabbableContentContainer = flow;
 
                 // if the piece belongs to a currently selected object, assume that the user wants to change all selected objects.
                 // if the piece belongs to an unselected object, operate on that object alone, independently of the selection.
-                var relevantObjects = (beatmap.SelectedHitObjects.Contains(hitObject) ? beatmap.SelectedHitObjects : hitObject.Yield()).ToArray();
-                var relevantSamples = relevantObjects.Select(h => h.Samples).ToArray();
+                relevantObjects = (beatmap.SelectedHitObjects.Contains(hitObject) ? beatmap.SelectedHitObjects : hitObject.Yield()).ToArray();
+                relevantSamples = relevantObjects.Select(GetSamples).ToArray();
 
                 // even if there are multiple objects selected, we can still display sample volume or bank if they all have the same value.
-                string? commonBank = getCommonBank(relevantSamples);
+                string? commonBank = getCommonBank();
                 if (!string.IsNullOrEmpty(commonBank))
                     bank.Current.Value = commonBank;
 
-                int? commonVolume = getCommonVolume(relevantSamples);
+                int? commonVolume = getCommonVolume();
                 if (commonVolume != null)
                     volume.Current.Value = commonVolume.Value;
 
-                updateBankPlaceholderText(relevantObjects);
+                updateBankPlaceholderText();
                 bank.Current.BindValueChanged(val =>
                 {
-                    updateBankFor(relevantObjects, val.NewValue);
-                    updateBankPlaceholderText(relevantObjects);
+                    updateBank(val.NewValue);
+                    updateBankPlaceholderText();
                 });
                 // on commit, ensure that the value is correct by sourcing it from the objects' samples again.
                 // this ensures that committing empty text causes a revert to the previous value.
-                bank.OnCommit += (_, _) => bank.Current.Value = getCommonBank(relevantSamples);
+                bank.OnCommit += (_, _) => bank.Current.Value = getCommonBank();
 
-                volume.Current.BindValueChanged(val => updateVolumeFor(relevantObjects, val.NewValue));
+                updateAdditionBankPlaceholderText();
+                updateAdditionBankText();
+                updateAdditionBankActivated();
+                additionBank.Current.BindValueChanged(val =>
+                {
+                    updateAdditionBank(val.NewValue);
+                    updateAdditionBankPlaceholderText();
+                });
+                additionBank.OnCommit += (_, _) => updateAdditionBankText();
```

Comment:
Can logic around here be simplified like this?

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs b/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
index 37a6fa8a22..f9e04a8ca7 100644
--- a/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
@@ -178,13 +178,12 @@ private void load()
                 // this ensures that committing empty text causes a revert to the previous value.
                 bank.OnCommit += (_, _) => bank.Current.Value = getCommonBank();
 
-                updateAdditionBankPlaceholderText();
                 updateAdditionBankText();
                 updateAdditionBankActivated();
                 additionBank.Current.BindValueChanged(val =>
                 {
                     updateAdditionBank(val.NewValue);
-                    updateAdditionBankPlaceholderText();
+                    updateAdditionBankText();
                 });
                 additionBank.OnCommit += (_, _) => updateAdditionBankText();
 
@@ -257,12 +256,6 @@ private void updateBankPlaceholderText()
                 bank.PlaceholderText = string.IsNullOrEmpty(commonBank) ? "(multiple)" : string.Empty;
             }
 
-            private void updateAdditionBankPlaceholderText()
-            {
-                string? commonAdditionBank = getCommonAdditionBank();
-                additionBank.PlaceholderText = string.IsNullOrEmpty(commonAdditionBank) ? "(multiple)" : string.Empty;
-            }
-
             private void updateAdditionBankActivated()
             {
                 bool anyAdditions = relevantSamples.Any(o => o.Any(s => s.Name != HitSampleInfo.HIT_NORMAL));
@@ -280,6 +273,7 @@ private void updateAdditionBankText()
                 if (string.IsNullOrEmpty(commonAdditionBank)) return;
 
                 additionBank.Current.Value = commonAdditionBank;
+                additionBank.PlaceholderText = string.IsNullOrEmpty(commonAdditionBank) ? "(multiple)" : string.Empty;
             }
 
             private void updateVolume(int? newVolume)
```

I have more to say about this code, but this would be a first step in the right direction.

Reply:
Updating the text in the `ValueChanged` event has a weird side-effect that it will never allow you to clear the textbox, because changes do not propagate when the textbox is empty. I'll find a way to simplify it in a different way

Reply:
Not much has simplified here. Speaking from personal experience (and I've made the same mistake myself), having multiple update methods like `updateAdditionBankText` and `updateAdditionBankVisual` and `updateAdditionBank` (with a bool) is not good news. Ideally you'd want just one `updateState` that gets everything correct.

Can you explain what the issues are here *before* I try and go in with a hammer to simplify this?

Reply:
I'll explain the issues as thoroughly as possible;

The update methods `updateBank`, `updateAdditionBank`, `updateVolume`, `addHitSample`, and `removeHitSample` must all only trigger when the user edits their respective value, because these are quite destructive in a sense that they synchronize the property across all selected hit objects.

Visuals must be updated before the bindings are created that update the hitobject properties on change, to prevent the hitobjects from being changed immediately when opening the popup.

Properties must be updated immediately on change, to allow the visuals in the timeline to reflect real-time changes.

Placeholder texts must be updated on change, to prevent it from incorrectly showing "(multiple)" after the user changes a property on multiple objects and clears the textbox.

The text of the textbox must not be updated on change, because otherwise the user will be unable to temporarily clear the textbox, because that is not a valid value.

Reply:
I think I managed to straighten this out by hoisting some null checks out of the set methods (see https://github.com/ppy/osu/pull/23443/commits/0efa028e0a82dc192303e2b99bead7ce1db66280). Please check whether this preserves all behaviours you wanted to, tests are passing as far as I could tell.

Reply:
Looks good. I did some small cleanup and fixed an unrelated bug i found while testing the behaviours.